File: Java/source/com/CxWave/Wave/Shell/ShellObjectManager.java
/*************************************************************************************************************************
 * Copyright (C) 2015-2016 Vidyasagara Guntaka & CxWave, Inc * All rights reserved. * Author : Vidyasagara Reddy Guntaka *
 *************************************************************************************************************************/

package com.CxWave.Wave.Shell;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;

import com.CxWave.Wave.Framework.Boot.BootCompleteForThisLocationEvent;
import com.CxWave.Wave.Framework.ObjectModel.WaveLocalObjectManagerForUserSpecificTasks;
import com.CxWave.Wave.Framework.ObjectModel.Annotations.ObjectManagerPriority;
import com.CxWave.Wave.Framework.ObjectModel.Boot.WaveAsynchronousContextForBootPhases;
import com.CxWave.Wave.Framework.Type.WaveServiceId;
import com.CxWave.Wave.Framework.Utils.Assert.WaveAssertUtils;
import com.CxWave.Wave.Framework.Utils.Source.WaveJavaSourceRepository;
import com.CxWave.Wave.Resources.ResourceEnums.ResourceId;
import com.CxWave.Wave.Resources.ResourceEnums.WaveObjectManagerPriority;

@ObjectManagerPriority (WaveObjectManagerPriority.WAVE_OBJECT_MANAGER_PRIORITY_SHELL)
public class ShellObjectManager extends WaveLocalObjectManagerForUserSpecificTasks
{
    private static ShellObjectManager s_shellObjectManager = null;

    private static String             s_shellRootPrompt    = "Wave";

    public ShellObjectManager ()
    {
        super (getServiceName ());

        setAllowAutomaticallyUnlistenForEvents (false);
    }

    public static String getServiceName ()
    {
        return ("Shell");
    }

    public static ShellObjectManager getInstance ()
    {
        if (null == s_shellObjectManager)
        {
            s_shellObjectManager = new ShellObjectManager ();
        }

        WaveAssertUtils.waveAssert (null != s_shellObjectManager);

        return (s_shellObjectManager);
    }

    public static WaveServiceId getWaveServiceId ()
    {
        return ((getInstance ()).getServiceId ());
    }

    @Override
    public void initialize (final WaveAsynchronousContextForBootPhases waveAsynchronousContextForBootPhases)
    {
        waveAsynchronousContextForBootPhases.setCompletionStatus (ResourceId.WAVE_MESSAGE_SUCCESS);
        waveAsynchronousContextForBootPhases.callback ();
    }

    @Override
    public void boot (final WaveAsynchronousContextForBootPhases waveAsynchronousContextForBootPhases)
    {
        waveAsynchronousContextForBootPhases.setCompletionStatus (ResourceId.WAVE_MESSAGE_SUCCESS);
        waveAsynchronousContextForBootPhases.callback ();
    }

    @Override
    public void listenForEvents (final WaveAsynchronousContextForBootPhases waveAsynchronousContextForBootPhases)
    {
        waveAsynchronousContextForBootPhases.setCompletionStatus (ResourceId.WAVE_MESSAGE_SUCCESS);
        waveAsynchronousContextForBootPhases.callback ();
    }

    private void bootCompleteForThisLocationEventHandler (final BootCompleteForThisLocationEvent bootCompleteForThisLocationEvent)
    {
        unlistenEvents ();

        infoTracePrintf ("ShellObjectManager.bootCompleteForThisLocationEventHandler : Entering ...");

        reply (bootCompleteForThisLocationEvent);

        // Get the Root Shell and instantiate it.

        final Class<?> shellRootClass = WaveJavaSourceRepository.getShellRootClass ();

        if (null != shellRootClass)
        {
            infoTracePrintf ("ShellObjectManager.bootCompleteForThisLocationEventHandler : Shell Root is handled by class : %s", shellRootClass.getTypeName ());
        }
        else
        {
            warnTracePrintf ("ShellObjectManager.bootCompleteForThisLocationEventHandler : Shell Root is not available and hence not spawning Shell.");

            return;
        }

        // Then instantiate and execute the root shell

        final Constructor<?>[] declaredConstructors = shellRootClass.getDeclaredConstructors ();
        Constructor<?> declaredConstructorWithNameParamerter = null;

        for (final Constructor<?> declaredConstructor : declaredConstructors)
        {
            if (1 == (declaredConstructor.getParameterCount ()))
            {
                declaredConstructorWithNameParamerter = declaredConstructor;

                break;
            }
        }

        if (null == declaredConstructorWithNameParamerter)
        {
            fatalTracePrintf ("ShellObjectManager.bootCompleteForThisLocationEventHandler : Shell Root clould not be instantiated sicne there is no constructor which takes name parameter.");
            waveAssert ();
        }

        Object shellObject = null;

        try
        {
            shellObject = declaredConstructorWithNameParamerter.newInstance (s_shellRootPrompt);
        }
        catch (InstantiationException | IllegalAccessException | IllegalArgumentException | InvocationTargetException e)
        {
            fatalTracePrintf ("ShellObjectManager.bootCompleteForThisLocationEventHandler : Could not instantiate shell root class : %s.  Details : %s", shellRootClass.getTypeName (), e.toString ());

            final Throwable cause = e.getCause ();

            if (null != cause)
            {
                fatalTracePrintf ("ShellObjectManager.bootCompleteForThisLocationEventHandler : Cause : %s:\n\r%s", cause.toString ());
            }

            waveAssert ();
        }

        final ShellBase shellBaseForRoot = (ShellBase) shellObject;

        shellBaseForRoot.shellExecuteHanler (null);
    }
}

File: C++/source/Shell/ShellObjectManager.cpp
/***************************************************************************
 *   Copyright (C) 2005 Vidyasagara Guntaka                                *
 *   All rights reserved.                                                  *
 *   Author : Amr Sabaa                                                    *
 ***************************************************************************/

#include "Shell/ShellObjectManager.h"
#include "Shell/ShellRegression.h"
#include "Framework/Utils/FrameworkToolKit.h"
#include "Framework/Utils/AssertUtils.h"
#include "Framework/Boot/BootCompleteForThisLocationEvent.h"
#include "Framework/Boot/BootTypes.h"
#include "Framework/Core/WaveFrameworkObjectManager.h"
#include "Framework/ObjectModel/WaveAsynchronousContextForBootPhases.h"

namespace WaveNs
{

ShellObjectManager::ShellObjectManager ()
    : WaveLocalObjectManagerForUserSpecificTasks ("Shell")
{
    setAllowAutomaticallyUnlistenForEvents (false);

    m_pWaveShell = new ShellWave (this);
}

ShellObjectManager *ShellObjectManager::getInstance ()
{
    static ShellObjectManager *pShellObjectManager = NULL;

    if (NULL == pShellObjectManager)
    {
        pShellObjectManager = new ShellObjectManager ();
        WaveNs::waveAssert (NULL != pShellObjectManager, __FILE__, __LINE__);
    }

    return (pShellObjectManager);
}

ShellObjectManager::~ShellObjectManager ()
{
    if (NULL != m_pWaveShell)
    {
        delete m_pWaveShell;
    }
}

WaveServiceId ShellObjectManager::getWaveServiceId ()
{
    return ((getInstance ())->getServiceId ());
}

void ShellObjectManager::initialize (WaveAsynchronousContextForBootPhases *pWaveAsynchronousContextForBootPhases)
{
    trace (TRACE_LEVEL_DEVEL, "ShellObjectManager::initialize : Entering ...");

    pWaveAsynchronousContextForBootPhases->setCompletionStatus (WAVE_MESSAGE_SUCCESS);
    pWaveAsynchronousContextForBootPhases->callback ();
}

void ShellObjectManager::boot (WaveAsynchronousContextForBootPhases *pWaveAsynchronousContextForBootPhases)
{
    trace (TRACE_LEVEL_DEVEL, "ShellObjectManager::boot : Entering ...");

    pWaveAsynchronousContextForBootPhases->setCompletionStatus (WAVE_MESSAGE_SUCCESS);
    pWaveAsynchronousContextForBootPhases->callback ();
}

void ShellObjectManager::listenForEvents (WaveAsynchronousContextForBootPhases *pWaveAsynchronousContextForBootPhases)
{
    listenForEvent (WaveFrameworkObjectManager::getWaveServiceId (), BOOT_COMPLETE_FOR_THIS_LOCATION, reinterpret_cast<WaveEventHandler> (&ShellObjectManager::bootCompleteForThisLocationEventHandler));

    pWaveAsynchronousContextForBootPhases->setCompletionStatus (WAVE_MESSAGE_SUCCESS);
    pWaveAsynchronousContextForBootPhases->callback ();
}

void ShellObjectManager::bootCompleteForThisLocationEventHandler (const BootCompleteForThisLocationEvent *&pBootCompleteForThisLocationEvent)
{
    // Since we told framework not to unlistenEvents, we must explicitly unlisten for events since we are going to go into an infinite while loop.

    unlistenEvents ();

    reply (reinterpret_cast<const WaveEvent *&> (pBootCompleteForThisLocationEvent));

    if (false == (FrameworkToolKit::getIsRunningAsADaemon ()))
    {
        trace (TRACE_LEVEL_INFO, "Now Entering Shell Mode.");

        m_pWaveShell->shellExecuteHandler ();
    }
}

}
